apollo-router/src/plugins/telemetry/mod.rs

```diff
@@ -256,116 +267,166 @@ impl Plugin for Telemetry {
+                })
+            }))
+            .map_future_with_request_data(
+                move |request: &router::Request| {
+                    if !use_legacy_request_span {
+                        let span = Span::current();
 
+                        span.set_dyn_attribute(
+                            HTTP_REQUEST_METHOD,
+                            request.router_request.method().to_string().into(),
+                        );
+                    }
 
-                async move {
-                    let span = Span::current();
-                    let response: Result<router::Response, BoxError> = fut.await;
+                    let client_name: &str = request
+                        .router_request
+                        .headers()
+                        .get(&config_request.apollo.client_name_header)
+                        .and_then(|h| h.to_str().ok())
+                        .unwrap_or("");
+                    let client_version = request
+                        .router_request
+                        .headers()
+                        .get(&config_request.apollo.client_version_header)
+                        .and_then(|h| h.to_str().ok())
+                        .unwrap_or("");
+
+                    let mut custom_attributes =
+                        config_request.spans.router.attributes.on_request(request);
+                    custom_attributes.extend([
+                        (CLIENT_NAME_KEY, client_name.to_string().into()),
+                        (CLIENT_VERSION_KEY, client_version.to_string().into()),
+                        (
+                            Key::from_static_str("apollo_private.http.request_headers"),
+                            filter_headers(
+                                request.router_request.headers(),
+                                &config_request.apollo.send_headers,
+                            )
+                            .into(),
+                        ),
+                    ]);
+
+                    custom_attributes
+                },
+                move |custom_attributes: HashMap<
+                    opentelemetry_api::Key,
+                    opentelemetry_api::Value,
+                >,
+                      fut| {
+                    let start = Instant::now();
+                    let config = config_later.clone();
 
-                    span.record(
-                        APOLLO_PRIVATE_DURATION_NS,
-                        start.elapsed().as_nanos() as i64,
-                    );
+                    Self::plugin_metrics(&config);
 
+                    async move {
+                        let span = Span::current();
+                        span.set_dyn_attributes(custom_attributes);
+                        let trace_id = TraceId::maybe_new()
+                            .map(|t| t.to_string())
+                            .unwrap_or_default();
+                        span.record("trace_id", trace_id);
```

Comment:
I think we can delete that part

Reply:
Hmm, not sure. For legacy mode the `trace_id` is still needed?

Reply:
Unless you added the functionality elsewhere?

Reply:
I deleted it because the trace_id is already given at the creation of the deprecated span in span_factory